docs/adrs/00009-conservative-purl-garbage-collection.md

```diff
@@ -0,0 +1,104 @@
+
+The challenge is determining which orphaned PURLs should be deleted during SBOM deletion and which should be preserved because they represent packages with security significance.
+
+### Assumptions
+
+* Remove orphaned PURL records when SBOMs are deleted to prevent unbounded database growth
+* Preserve PURL records for packages that have security advisories referencing them
+* Maintain referential integrity across the three-tier PURL hierarchy
+
+### Technical Constraints
+
+* **Transaction ordering**: GC must execute BEFORE the SBOM delete statement, as SBOM deletion triggers CASCADE deletes on `sbom_package_purl_ref` which the GC query relies on
+* **Foreign key relationships**: The database enforces CASCADE constraints from `base_purl` → `versioned_purl` → `qualified_purl`
+* **Advisory preservation**: Any `base_purl` referenced in `purl_status` must be preserved along with ALL its versions and qualifiers
+
+## Decision
+
+Implement a **conservative garbage collection approach** that preserves all versions and qualifiers of a package if ANY advisory references its base PURL.
+
+The GC process:
+
+1. Identifies `qualified_purl` records that will become orphaned after SBOM deletion by checking `sbom_package_purl_ref`
+2. Traces from `qualified_purl` → `versioned_purl` → `base_purl` to identify the base package
+3. **Conservatively excludes** ANY `base_purl` that has entries in `purl_status` (advisory references)
+4. Deletes only the `qualified_purl` records that are orphaned AND whose base package has no advisory references
+
+### Conservative Nature
+
+The approach is "conservative" because:
+
+* If `pkg:maven/io.quarkus/quarkus-core` (base) is referenced by ANY advisory
+* Then ALL versions are preserved: `@3.2.11`, `@3.2.12`, `@3.3.0`, etc.
+* And ALL qualifiers are preserved: `?type=jar`, `?arch=x86_64`, etc.
+
+All versions are preserved because it cannot be determined the versioned and qualified PURLs ingested during advisory ingestion.
+
+## Alternative Approaches
+
+### Deferred/Batch GC
+
+Separate GC from SBOM deletion, running it as a periodic background job:
+
+**Pros:**
+* Simpler SBOM deletion endpoint logic
+* Could optimize GC across multiple deletions
+
+**Cons:**
+* Temporary database bloat between deletions and GC runs
+* More complex orchestration and monitoring
+* Risk of orphaned records persisting indefinitely if GC job fails
+* Transaction boundaries become unclear - harder to reason about consistency
+
+### No GC - Keep All PURLs
+
+Never delete PURL records, treating them as a cumulative knowledge base:
+
+**Pros:**
+* Simplest implementation
+* Complete historical record
+* No risk of data loss
+
+**Cons:**
+* Unbounded database growth
+* Performance degradation over time
+* Violates user expectations for data deletion
```

Comment:
I wonder if this is true. Users never explicitly create packages, so I'm not sure they'd have any expectations for deleting them. Packages just magically appear when either an advisory or an sbom is ingested.